### ISSUE TITLE
How to compile it at openwrt 14.07?

### ISSUE THREAD

**Issue by sgon00** [trojan-gfw/openwrt-trojan#27]
Hi, my device has no way to upgrade to modern openwrt versions. Vendor doesn't support it.
It's running openwrt 14.07. 

I have tried many ways for a week to compile trojan, but no luck. 

I tried to upgrade openwrt, but failed many times. Tried to upgrade to 15, 17, 18, 19 with different dts files and .config files etc..., but no luck.

I tried to compile trojan on openwert 14.07 at debian7, debian8, debian10. 

Hack the code etc.. There are many stories. I even combile half at debian7 and then copy the half work to debian10 and continue the compilation, this is because packages requirement for 14.07 official release and trojan are different. etc.. etc..

In conclusion, I don't find a way to complete this job. 

Can anyone help me on this? 

I don't need any cutting-edge features such as tls13 on openssl1.1. As long as it work, it should be fine. I also tried to comment out its own openssl1.1 related lines in Makefile and add `+libopenssl` on Depends trying to use the built-in openssl. (not sure if I did it correctly since there is no doc on this), but again, all attempts failed. 

Thank you very much for your help.

**Comment by aon3ko** [comment]
Paste your build log on or we can't help you.

**Comment by sgon00** [comment]
@EXSERENS Thank you very much for the reply. As what I said, it has many stories on this compilation. One build.log will not really explian them all. Let me explain some problems I had from memory.

(1) All packages from official openwrt 14.07 require cmake2.x and old host glibc etc.. For exampe, boost and its all dependencies. Thus, they have to be compile on debian7 or debian8.

(2) trojan require cmake3.x to compile. It does not exist in openwrt 14.07. It also requires latest host glibc. What I did here is compile boost and its dependencies which require cmake2.x and other old libs on debian7/debian8. Tar the source, scp transfer to debian10, and continue the compilation for trojan. This is really a hack and I really shouldn't do this and I have no ideas if it will work even if the compiliation will complete. But it won't. 

(3) trojan requires boost 1.66 and above. But openwrt 14.07 boost is about 1.55. So that is another failure. I did overcome this, by hacking the requirement code file (by hard coding the required version to be 1.55+ instead of 1.66+). But obviously this won't work. This is a contradiction. I don't think boost 1.66 and above will work on openwrt 14.07 (not even sure how to compile boost 1.66 for openwrt 14.07), and I don't think trojan can be compiled successfully with the hack and boost 1.55. 

(4) even if I remove all openssl1.1 related lines from trojan Makefile, and even if I compiled openssl1.1 successfully with many many hack, trojan still report it's looking for 1.1 somehow. Well, I hacked this again. I forgot what I did. I copied/pasted the openssl1.1 compiled result to some directories to overcome this error.

(5) About using cmake3.x on openwrt14.07 at debian10 and compile 14.07 at debian10, that is another long story. openwrt14.07 doesn't support modern debian compilation at all. i have to first compile openwrt 18.06, and then copy many executables from 18.06 compilation result such as `staging_dir/host/bin/mkhash`, `cmake` version 3 and its share files etc.. from 18.06 compiation to 14.07 compilation to workaround many compile errors on openwrt 14.07 at debian10. I also need to self modify many files too, such as `./scripts/feeds` does not support debian10 env, `staging_dir/host/bin/ipkg-build` can not read trojan description etc.. (btw, ipkg-build is used on openwrt14.07 and does not exist on latest openwrt). These files are just too old to be used for trojan and modern debian.

(6) Finally, after many many hacks which I shouldn't do and did blindly, you can imagine that trojan compile finally failed with cpp errors, something is not declared etc... I don't blame this error. Because all steps I did are just crazy and don't make sense. 

The above problems are just some of the nightmare. There were more workarounds I did for trojan compilation ends with cpp not declare errors. I really hope trojan can provide a way to compile at 14.07. I really want to use it. 

Btw, regarding upgrading the device openwrt system to modern openwrt solution, that is another long hacking story you never want to know. More crazy than this compilation..

Btw, I have removed the hacked compilation directory because I feel that doesn't really make sense and doubt trojan will run even if I complete the compilation after all the hacks. If you think that might run, I can redo all the hacks and paste the final error I met (which I don't think the error msg will make any sense to anyone). 

Thank you very much for your help agian.

**Comment by wongsyrone** [comment]
Sorry, you must meet all the dependencies, if the official repo doesn't offer the package, please modify the package Makefile to compile on the extremely ancient platforms.
